packages/ts/react-routing/src/index.tsx

```diff
@@ -0,0 +1,32 @@
+import { type IndexRouteObject, type NonIndexRouteObject, useMatches } from 'react-router-dom';
+
+export type MenuProps = Readonly<{
+  icon?: string;
+  title?: string;
+}>;
+
+export type AccessProps = Readonly<{
+  requireAuthentication?: boolean;
+  rolesAllowed?: readonly string[];
+}>;
+
+export type RouteMetadata = AccessProps & MenuProps;
+
+type HandleWithMetadata = { handle?: RouteMetadata };
+
+type Override<T, E> = E & Omit<T, keyof E>;
+
+export type IndexRouteObjectWithMetadata = Override<IndexRouteObject, HandleWithMetadata>;
+export type NonIndexRouteObjectWithMetadata = Override<
+  Override<NonIndexRouteObject, HandleWithMetadata>,
+  {
+    children?: RouteObjectWithMetadata[];
+  }
+>;
+export type RouteObjectWithMetadata = IndexRouteObjectWithMetadata | NonIndexRouteObjectWithMetadata;
```

Comment:
One downside with this type is that developers would not be able to add any other custom property to `handle` anymore. A future improvement could make it generic, otherwise I'd suggest to keep it simple for now.